hail/python/hailtop/aiotools/fs.py

```diff
@@ -281,6 +282,80 @@ def __init__(self, src: Union[str, List[str]], dest: str, *, treat_dest_as: str
         self.treat_dest_as = treat_dest_as
 
 
+class SourceReport:
+    def __init__(self, source):
+        self._source = source
+        self._source_type: Optional[str] = None
+        self._files = 0
+        self._errors = 0
+        self._complete = 0
+        self._first_file_error: Optional[Dict[str, Any]] = None
+        self._exception: Optional[Exception] = None
+
+    def set_exception(self, exception: Exception):
+        assert not self._exception
+        self._exception = exception
+
+    def set_file_error(self, srcfile: str, destfile: str, exception: Exception):
+        if self._first_file_error is None:
+            self._first_file_error = {
+                'srcfile': srcfile,
+                'destfile': destfile,
+                'exception': exception
+            }
+
+    def raise_first_exception(self):
+        if self._exception:
+            raise self._exception
+        if self._first_file_error:
+            raise self._first_file_error['exception']
+
+
+class TransferReport:
+    def __init__(self, transfer: Transfer):
+        self._transfer = transfer
+        if isinstance(transfer.src, str):
+            self._source_report = SourceReport(transfer.src)
+        else:
+            self._source_report = [SourceReport(s) for s in transfer.src]
+        self._exception: Optional[Exception] = None
+
+    def set_exception(self, exception: Exception):
+        assert not self._exception
```

Comment:
I assume in the future these will store the number of bytes transferred and the time it took?

Reply:
Yep!